### PR TITLE
refactor(core): virtualize execution rendering

### PR DESCRIPTION
**Pull request by anotherchrisberry**
I went into a rabbit hole that looked small (can we virtualize the executions view) and I found it was full of rats nests, which, well, it happens. A lot of this code has been migrated from Angular to React bit by bit, and there were a ton of optimizations in place for an Angular world (mutating objects in place to minimize DOM re-renders) that are really bad for React.

Overall, the performance is pretty good. We're able to handle much larger quantities of executions and much larger apps (e.g. apps with 50-100 different pipelines) without having to rely on the user collapsing the ones they don't want to look at right this second.

### killed rats
* we were setting a `reloadingForFilters` flag on the executions data source and using that to block the view when the user changed the number of executions per pipeline or the pipelines via the side filter. This was just incidentally picked up on re-render of `<Executions>` because there were like 40 other mutations that happened along the way, and, through some small miracle, it always happens to be reset. I moved it to a state field, and passed a function to the <ExecutionFilters> as a prop to allow it to toggle it.
* `forceUpdate` calls are out. These were there because of the Angular-style mutations of elements that happened whenever the execution groups or the execution data changed.

### Re-trained rats
I've done some light touch tweaks to treat objects a little more like they're immutable, specifically around the `executions` data: 
  * replace the execution in the data source when it changes
  * replace the execution group on the `ExecutionFilterModel` if its data changes

### Rats I tried to re-train, but really they re-trained me along the way
* `element.scrollIntoView` is a thing that browsers let you do now, which is really cool! It also is...kind of a pain to make sure the element is visible first, especially once you start virtualizing things. I took a hacky-ish route out and just forced the deep linked thing to be rendered all the time.
* There's a state object in one of those views that's an object, and we do some nasty `this.state.sortFilter.showDurations = ...` stuff in a few places. I thought for sure it would be easy to refactor and it wasn't.

### Rat King
* routing is Rat King and routing is still the worst. I tried to do some refactoring around initializing the model and fixing browser navigation, thinking I could make the back button work, ~but I gave up. Another day.~ and I think it's working as expected now (which means I broke something else). You can use the browser back/forward buttons and it should change filters without going crazy celery.

### CODE REVIEW

File: app/scripts/modules/core/src/utils/RenderWhenVisible.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      {
+        root: container,
+        threshold: [0],
+        rootMargin: bufferHeight && container ? `${bufferHeight}px 0px` : '0px',
+      },
+    );
+    observer.observe(node);
+  }, []);
+
+  return (
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/utils/RenderWhenVisible.tsx]: Another way to do this is with refs and deps:

```
function Node() {
    const nodeRef = useRef();

    useEffect(() => {
      if (!nodeRef.current) {
        return;
      }

      console.log('Register observer')
      return () => console.log('Unregister observer')
    }, [nodeRef.current])
    return (
      <div ref={nodeRef}>
        Here's a node
      </div>
    );
}
```

File: app/scripts/modules/core/src/pipeline/executions/Executions.tsx [hunk 2]
```diff
@@ -287,6 +259,7 @@ export class Executions extends React.Component<IExecutionsProps, IExecutionsSta
   private showCountChanged = (event: React.ChangeEvent<HTMLSelectElement>): void => {
     const value = event.target.value;
     this.state.sortFilter.count = parseInt(value, 10);
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/executions/Executions.tsx]: turns out it's super necessary unless we want to refactor all the view state controls!

(the line below that I added is _not_ necessary and will be removed now)

File: app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroup.tsx [hunk 3]
```diff
@@ -220,7 +225,7 @@ export class ExecutionGroup extends React.Component<IExecutionGroupProps, IExecu
     if (group.targetAccounts && group.targetAccounts.length > ACCOUNT_TAG_OVERFLOW_LIMIT) {
       groupTargetAccountLabels.push(
         <span
-          key="foo"
+          key="account-overflow"
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroup.tsx]: lol

File: app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx [hunk 4]
```diff
@@ -47,12 +53,20 @@ export class ExecutionGroups extends React.Component<IExecutionGroupsProps, IExe
   }
 
   private showingDetails(): boolean {
+    if (this.state) {
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx]: good question - it was called in the constructor; however, I simplified the state initialization and I think it's cleaner now.

File: app/scripts/modules/core/src/utils/RenderWhenVisible.tsx (hunk 1 above)
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/utils/RenderWhenVisible.tsx]: Updated to just `useEffect` with a `ref`. I was surprised in my testing to see I didn't need a null check on `nodeRef.current` - it's always there - and I didn't need to include `nodeRef.current` in my `deps`. The whole thing ended up a little cleaner.

File: app/scripts/modules/core/src/utils/RenderWhenVisible.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+export const RenderWhenVisible = ({
+  bufferHeight = 0,
+  initiallyVisible,
+  placeholderHeight,
+  render,
+  container,
+  disableHide,
+}: IRenderWhenVisibleProps) => {
+  const [isVisible, setIsVisible] = React.useState(!!initiallyVisible);
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/utils/RenderWhenVisible.tsx]: We like hooks now.

File: app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx [hunk 6]
```diff
@@ -47,12 +53,20 @@ export class ExecutionGroups extends React.Component<IExecutionGroupsProps, IExe
...
+      }
+    }
     return ReactInjector.$state.includes('**.execution');
   }
 
-  public shouldComponentUpdate(_nextProps_: IExecutionGroupsProps, nextState: IExecutionGroupsState): boolean {
-    return nextState.groups !== this.state.groups || nextState.showingDetails !== this.state.showingDetails;
-  }
+  private setContainer = (container: HTMLDivElement) => {
+    if (this.state.container !== container) {
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx]: The problem with `createRef` is that we need the ref assignment to trigger an update, i.e. when `containerRef.current` changes, that is when we need to render the child component. I tried passing the `React.RefObject` down to the child, as well as sending `containerRef.current` as a prop, but it doesn't re-render, so you end up with nothing. 

The state approach is maybe a little hacky, but I couldn't find a cleaner way to do it.

File: app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx [hunk 7]
```diff
@@ -47,12 +53,20 @@ export class ExecutionGroups extends React.Component<IExecutionGroupsProps, IExe
   }
 
   private showingDetails(): boolean {
+    if (this.state) {
+      const { executionId } = ReactInjector.$stateParams;
+      if (this.state.groups.every(g => g.executions.every(e => e.id !== executionId))) {
```
[the remark this answers is not in the document]
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/executions/executionGroup/ExecutionGroups.tsx]: Good point, added comments

**Pull request merged**